### PR TITLE
Ignore key presses when modifier is held but shortcut has no modifier defined

### PR DESCRIPTION
**Pull request by taras-yemets**
Previously if the shortcut had no held modifier key defined, is was matched regardless of whether any modifier key was held.
This logic had an undesired side-effect: depending on the shortcut, the default browser/OS behavior was overridden or tweaked. For example, https://github.com/sourcegraph/sourcegraph/pull/46829 added `C` key shortcut to focus the code editor on the blob page. Selecting some text outside the code editor and pressing `Cmd+C` on Mac not only copied text to the clipboard (expected behavior) but also focused the code editor (unexpected part).
This PR changes the ShortcutManager logic to match shortcuts with no modifiers defined only if no modifier keys were pressed along with the shortcut key.

## Test plan
- CI passes
- Shortcuts with no modifier keys defined should not be matched if any modifier key was pressed along with the shortcut key. For example: 
  - ensure `blob-page-switch-areas-shortcuts` feature flag is enabled
  - select any text outside the code editor and press `Cmd+C` - text should be copied, and the active element shouldn't change
  - `C` key press (without a modifier) should focus the code editor

## App preview:

- [Web](https://sg-web-taras-yemets-fix-focus-blob.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by fkling** [comment]
Makes me wonder whether we should change the shortcut handler's implementation to match shortcuts with no modifier only when no modifier is held. I feel like that aligns more with the expectation.

File: client/web/src/repo/blob/CodeMirrorBlob.tsx [hunk 1]
```diff
@@ -437,17 +437,21 @@ export const CodeMirrorBlob: React.FunctionComponent<BlobProps> = props => {
...
-                        {...keybinding}
-                        allowDefault={true}
-                        onMatch={() => {
-                            editorRef.current?.contentDOM.focus()
-                        }}
-                    />
-                ))}
+            {enableBlobPageSwitchAreasShortcuts && focusCodeEditorShortcut && (
+                <>
+                    <Shortcut ordered={['c']} held={['Mod']} allowDefault={true} onMatch={noop} />
```
Comment (fkling) [inline client/web/src/repo/blob/CodeMirrorBlob.tsx:442]: Aside from possibly fixing this inside the shortcut manager itself, the problem I see with this implementation is that if we ever change the real shortcut we'd also have to remember to update this one here.

**Comment by taras-yemets** [comment]
> Makes me wonder whether we should change the shortcut handler's implementation to match shortcuts with no modifier only when no modifier is held. I feel like that aligns more with the expectation.

@fkling, makes sense to me! What do you think about this implementation https://github.com/sourcegraph/sourcegraph/pull/49055/commits/8f2bb9069b2dd345f52f2621bc62120059efded6?

File: client/shared/src/react-shortcuts/ShortcutManager.test.tsx [hunk 2]
```diff
@@ -273,5 +273,20 @@ describe('ShortcutManager', () => {
...
+        it("doesn't match shortcut when any modifier key is held, but no modifier key is defined for the shortcut", () => {
+            const fooSpy = sinon.spy()
+
+            render(
+                <ShortcutProvider>
+                    <Shortcut ordered={['/']} onMatch={fooSpy} />
+                </ShortcutProvider>
+            )
+
+            for (const key of ['Alt', 'Control', 'Meta', 'Shift']) {
```
Comment (taras-yemets) [inline client/shared/src/react-shortcuts/ShortcutManager.test.tsx:286]: For some reason rest of [`MODIFIER_KEYS`](https://github.com/sourcegraph/sourcegraph/pull/49055/files#diff-928492885d0a0b94b0a046decb40289f29c592a5b8b3daa5b97e7b6151d1ef93R78) don't work correctly with `userEvent.keyboard`: the event is fired in the test.

**Comment by fkling** [comment]
Nice! I'd like to know whether the rest of  @sourcegraph/code-exploration-devs agrees with this approach.
You need to update your PR description ;)

**Comment by philipp-spiess** [comment]
> Makes me wonder whether we should change the shortcut handler's implementation to match shortcuts with no modifier only when no modifier is held. I feel like that aligns more with the expectation.

YES!

File: client/shared/src/react-shortcuts/ShortcutManager.test.tsx [hunk 3]
```diff
@@ -273,5 +273,20 @@ describe('ShortcutManager', () => {
 
             platform.reset()
         })
+
+        it("doesn't match shortcut when any modifier key is held, but no modifier key is defined for the shortcut", () => {
```
Comment (philipp-spiess) [inline client/shared/src/react-shortcuts/ShortcutManager.test.tsx:277]: ❤️

File: client/shared/src/react-shortcuts/keys.ts [hunk 4]
```diff
@@ -75,21 +75,24 @@ export type AlphabetKey =
...
+    'FnLock',
+    'Hyper',
+    'Meta',
+    'NumLock',
+    'ScrollLock',
+    'Shift',
+    'Super',
+    'Symbol',
+    'SymbolLock',
+] as const
```
Comment (philipp-spiess) [inline client/shared/src/react-shortcuts/keys.ts:93]: Wow so many keys here I never heard of 😱

**Review by philipp-spiess — APPROVED** [review]

File: client/shared/src/react-shortcuts/keys.ts (hunk 4 above)
Comment (taras-yemets) [reply client/shared/src/react-shortcuts/keys.ts:93]: I believe this list comes from [this table](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState#modifier_keys_on_gecko) on the MDN `KeyboardEvent.getModifierState()` page. 
[Not all of them are supported in tests](https://github.com/sourcegraph/sourcegraph/pull/49055/files#r1132151704), but I decided to check all of them in [`isAnyModifierKeyHeld`](https://github.com/sourcegraph/sourcegraph/pull/49055/files#diff-b7247afe28c4a747618dbb9ef3016a2ebd5cb6936cf94e95adc72de54e263dd1R123) for consistency.

**Review by fkling — APPROVED** [review]

**Pull request merged**